src/base/static/components/app.js

```diff
@@ -159,81 +266,255 @@ class App extends Component {
+                            datasetDownloadConfig={
+                              this.props.appConfig.dataset_download
+                            }
+                            apiRoot={this.props.appConfig.api_root}
+                          />
+                        </Suspense>
+                      );
+                    }}
+                  />
+                  <Route
+                    exact
+                    path="/page/:pageSlug"
+                    render={props => {
+                      if (
+                        !this.props.pageExists(
+                          props.match.params.pageSlug,
+                          Mapseed.languageCode,
+                        )
+                      ) {
+                        window.history.pushState("", "", "/");
+                        return (
+                          <Suspense fallback={<Fallback />}>
+                            <MapTemplate
+                              uiConfiguration="map"
+                              languageCode={Mapseed.languageCode}
+                              {...props.match}
+                            />
+                          </Suspense>
+                        );
+                      }
+
+                      return (
+                        <Suspense fallback={<Fallback />}>
+                          <MapTemplate
+                            uiConfiguration="customPage"
+                            languageCode={Mapseed.languageCode}
+                            {...props.match}
+                          />
+                        </Suspense>
+                      );
+                    }}
+                  />
+                  <Route
+                    exact
+                    path="/:datasetClientSlug/:placeId"
+                    render={props => {
+                      return (
+                        <Suspense fallback={<Fallback />}>
+                          <MapTemplate
+                            uiConfiguration="placeDetail"
+                            languageCode={Mapseed.languageCode}
+                            {...props.match}
+                          />
+                        </Suspense>
+                      );
+                    }}
+                  />
+                  <Route
+                    exact
+                    path="/:datasetClientSlug/:placeId/response/:responseId"
+                    render={props => {
+                      return (
+                        <Suspense fallback={<Fallback />}>
+                          <MapTemplate
+                            uiConfiguration="placeDetail"
+                            languageCode={Mapseed.languageCode}
+                            {...props.match}
+                          />
+                        </Suspense>
+                      );
+                    }}
+                  />
+                  <Route
+                    render={props => {
+                      window.history.pushState("", "", "/");
```

Comment:
Yeah, this was a little weird. We can't do a react router history update in a `render` method because it throws React into an endless update cycle.

Maybe instead we could have another ui "configuration" in `MainMap` called 'invalidRoute` or something, which handles the route back to `/`.

Do you think that makes sense?

Reply:
I'm not too familiar with the best pattern to use here, but I wonder if this can be solved with the `history` api: https://reacttraining.com/react-router/web/api/history

And the `history` package: https://github.com/ReactTraining/history#history--

Reply:
I believe using the history API here sends React into an update loop, since we would be updating state from within a `render` method.

I think we just need to move this router call to `MainMap`.